src/file.rs

```diff
@@ -193,26 +138,36 @@ impl TableFile {
 		capacity += GROW_SIZE_BYTES / entry_size as u64;
 
 		self.capacity.store(capacity, Ordering::Relaxed);
-		let mut file = self.file.upgradable_read();
-		if file.is_none() {
-			let mut wfile = RwLockUpgradableReadGuard::upgrade(file);
-			*wfile = Some(self.create_file()?);
-			file = RwLockWriteGuard::downgrade_to_upgradable(wfile);
+		let mut map_and_file = self.map.write();
+		match map_and_file.as_mut() {
+			None => {
+				let file = self.create_file()?;
+				try_io!(file.set_len(capacity * entry_size as u64));
+				let mut map = try_io!(unsafe { memmap2::MmapMut::map_mut(&file) });
+				madvise_random(&mut map);
+				*map_and_file = Some((map, file));
+			},
+			Some((map, file)) => {
+				try_io!(file.set_len(capacity * entry_size as u64));
+				let mut m = try_io!(unsafe { memmap2::MmapMut::map_mut(&*file) });
```

Comment:
Do we need to drop the existing mmap before creating this one?
(looks good if it works fine this way)

Reply:
That's probably a good idea